File: GodotKotlin/MultiplatformFrogGame/src/main/kotlin/com/example/game/cst/behavior/LFA.java
package com.example.game.cst.behavior;

import com.example.game.cst.behavior.ValueBasedRL;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Set;

public class LFA extends ValueBasedRL {
    private LinkedHashMap<String, Double> weights = new LinkedHashMap<String, Double>();
    private FeaturesExtractor extractor; // new featuresExtractor();

    public LFA(Double alpha, Double gamma, Integer numActions, FeaturesExtractor fe) {
        super(alpha, gamma, numActions);
        this.extractor = fe;
    }

    @Override
    public void update(ArrayList<Domain> state, ArrayList<Domain> newState,
                       Domain action, Domain reward) {
        LinkedHashMap<String, Double> gradient = this.extractor.getFeatures(state, action);

        Double new_q_val = this.getBestValue(newState);
        Double q_val = this.getValue(state, action);

        Double target = reward.doubleValue() + this.GAMMA * new_q_val;

        Set<String> name_features = gradient.keySet();
        for(String f : name_features) {
            Double w = this.ALPHA * (target - q_val) * gradient.get(f);
            this.weights.put(f, w);
        }
    }

    public LinkedHashMap<String, Double> getWeights() {
        return this.weights;
    }

    @Override
    protected Double getValue (ArrayList<Domain> state, Domain action) {
        Double q_val = (double) 0;
        LinkedHashMap<String, Double> gradient = this.extractor.getFeatures(state, action);

        Set<String> features = gradient.keySet();
        for (String f : features) {
            q_val += this.weights.get(f)*gradient.get(f);
        }

        return q_val;
    }

    @Override
    protected ArrayList<Double> getValues(ArrayList<Domain> state) {
        ArrayList<Double> vals = null;
        for (Integer a = 0; a < this.numActions; a++) {
            vals.add(getValue(state, new Domain(a)));
        }
        return vals;
    }


}

File: GodotKotlin/MultiplatformFrogGame/src/main/kotlin/com/example/game/cst/behavior/LFA.kt
package com.example.game.cst.behavior

class LFA() {

/*
import java.util.ArrayList
import java.util.Collections
import java.util.LinkedHashMap
import java.util.Set
import com.example.game.cst.RLKotlin.FeaturesExtractor

class LFA(alpha: Double, gamma: Double, epsilon: Double, epsilonDecay: Double, finalEpsilon: Double,
          numActions: Integer?, fe: FeaturesExtractor) :
    ValueBasedRL(alpha, gamma, epsilon: Double, epsilonDecay: Double, finalEpsilon: Double, numActions) {

    private val weights: LinkedHashMap<String, Double> = LinkedHashMap<String, Double>()
    private val extractor // new FeaturesExtractor();
            : FeaturesExtractor

    class Extactor() {
        abstract fun getFeatures(state: ArrayList<Domain?>?, action: Integer?): LinkedHashMap<String, Double> {}
    }

    abstract fun getFeatures(state: ArrayList<Domain?>?, action: Integer?): LinkedHashMap<String, Double> {}

    init {
        extractor = fe
    }

    @Override
    fun init(env: Environment) {
        val gradient: LinkedHashMap<String, Double> = extractor.getFeatures(env.getObservationSpace(), 0)
        val features: Set<String> = gradient.keySet()
        for (f in features) {
            weights.put(f, Math.random())
        }
    }

    @Override
    fun update(lastState: ArrayList<Domain?>?, state: ArrayList<Domain?>?,
               action: Integer?, reward: Double) {
        val gradient: LinkedHashMap<String, Double> = extractor.getFeatures(lastState, action)
        val new_q_val: Double = this.getBestValue(state)
        val q_val = getValue(lastState, action)
        val target: Double = reward + this.GAMMA * new_q_val
        val name_features: Set<String> = gradient.keySet()

        for (f in name_features) {
            val w: Double = this.ALPHA * (target - q_val) * gradient.get(f)
            weights.put(f, w)
        }
    }

    fun getWeights(): LinkedHashMap<String, Double> {
        return weights
    }

    @Override
    protected operator fun getValue(state: ArrayList<Domain?>?, action: Integer?): Double {
        var q_val = 0.0
        val gradient: LinkedHashMap<String, Double> = extractor.getFeatures(state, action)
        val features: Set<String> = gradient.keySet()

        for (f in features) {
            q_val += weights.get(f) * gradient.get(f)
        }
        return q_val
    }

    @Override
    protected fun getValues(state: ArrayList<Domain?>?): ArrayList<Double>? {
        val vals: ArrayList<Double>? = null

        for (a in 0 until this.numActions) {
            vals.add(getValue(state, a))
        }
        return vals
    }

    @Override
    protected fun proc() {
        var lastState : ArrayList<Domain>?
        var idAction = super.epsilonGreedyPolicy(
            this.epsilon, state
        )
        lastState = state

        state = stateMO?.i as ArrayList<Domain>?
        reward = rewardMO?.i as Double
        info = infoMO?.i as String
        done = doneMO?.i as Boolean

        episodeReward += reward

        this.update(lastState, state, idAction, reward)
        if (done)
            episodeReward = 0
            super.updateEpsilon()
    }

//    @Override
//    protected fun accessMemoryObjects() {
//    }

//    @Override
//    protected fun calculateActivation() {
//    }

    */
}

